packages/runtime-common/realm-auth-handler.ts

```diff
@@ -0,0 +1,125 @@
+
+export class RealmAuthHandler {
+  // Cached realm info to avoid fetching it multiple times for the same realm
+  private visitedRealms = new Map<
+    string,
+    {
+      isPublicReadable: boolean;
+      realmAuthClient?: RealmAuthClient;
+      url: string;
+    }
+  >();
+  private matrixClient: MatrixClient;
+  private loader: Loader;
+  private realmURL: string;
+
+  constructor(matrixClient: MatrixClient, loader: Loader, realmURL: string) {
+    this.matrixClient = matrixClient;
+    this.loader = loader;
+    this.realmURL = realmURL;
+  }
+
+  // A separate method for realm auth client creation to support mocking in tests
+  private createRealmAuthClient(
+    realmURL: URL,
+    matrixClient: MatrixClient,
+    loader: Loader,
+  ) {
+    return new RealmAuthClient(realmURL, matrixClient, loader);
+  }
+
+  fetchWithAuth = async (
+    request: Request,
+    retryOnAuthFail = true,
+  ): Promise<Response | null> => {
+    // todo rename method and retryOnAuthFail
+    if (request.url.startsWith(PACKAGES_FAKE_ORIGIN)) {
+      return null;
+    }
+
+    if (
+      request.url.endsWith('_session') ||
+      request.method === 'HEAD' ||
+      request.headers.has('Authorization')
+    ) {
+      return null;
+    }
+
+    let targetRealm: {
+      isPublicReadable: boolean;
+      realmAuthClient?: RealmAuthClient;
+      url: string;
+    };
+
+    let visitedRealmURL = Array.from(this.visitedRealms.keys()).find((key) => {
+      return request.url.includes(key);
+    });
+
+    if (visitedRealmURL) {
+      targetRealm = this.visitedRealms.get(visitedRealmURL)!;
+    } else {
+      let targetRealmHeadResponse = await this.loader.fetch(request, {
+        method: 'HEAD',
+      });
+
+      let targetRealmURL =
+        targetRealmHeadResponse.headers.get('x-boxel-realm-url');
+
+      if (!targetRealmURL) {
+        return null; // It doesn't look like we are talking to a realm (the request is for something else), so skip adding auth header
+      }
+
+      let isPublicReadable = Boolean(
+        targetRealmHeadResponse.headers.get('x-boxel-realm-public-readable'),
+      );
+
+      targetRealm = {
+        isPublicReadable,
+        url: targetRealmURL,
+        realmAuthClient: this.createRealmAuthClient(
+          new URL(targetRealmURL),
+          this.matrixClient,
+          this.loader,
+        ),
+      };
+
+      this.visitedRealms.set(targetRealmURL, targetRealm);
+    }
+
+    if (!targetRealm || !targetRealm.realmAuthClient) {
+      throw new Error(
+        `bug: should not have been able to get here without a visitedRealm or without an auth client`,
+      );
+    }
+
+    let isRequestToItself = targetRealm.url === this.realmURL; // Could be a request to itself when indexing its own cards
+
+    if (
+      isRequestToItself ||
+      (targetRealm.isPublicReadable && request.method === 'GET')
+    ) {
+      return null; // No need to add auth for GET to public readable realms
```

Comment:
This comment is confusing bcos its also handling request to itself

Reply:
Yeah I removed it because it's not really needed